### ISSUE TITLE
If the positions "x_coor" and "y_coor" should be swapped in Line 466 and 468 det3d/models/necks/rpn_transformer?

### ISSUE THREAD

**Issue by Castiel-Lee** [TuSimple/centerformer#25]
![image](https://user-images.githubusercontent.com/91303830/220546657-3366f194-33bd-4155-ba18-3afe96885dab.png)

Those two lines seem flawless.
But coincidently, I changed the range of PC, which made H unequal to W.  I encountered this:
![image](https://user-images.githubusercontent.com/91303830/220546893-24fa0d3c-7a19-4eb7-95fa-43321df0603e.png)
Afte debugging, in ```get_multi_scale_feature```, _center_pos_ sometimes fell out of the range of _feat_. After checking backwards, I found some elements of _y_coor_ > W exist. in Line 468.  
After some experiments I tried, this problem can be fixed by swaping _x_coord_ and _y_coord_.

**Comment by edwardzhou130** [comment]
Thanks for reporting this error. I think the position is correct. I generate the index in the 2D matrix in this way: https://github.com/TuSimple/centerformer/blob/5a949b88ed7bb15aafb39bf78c95f1452063ebea/det3d/models/necks/rpn_transformer.py#L247-L250
Here ``neighbor_coords[:, :, :, 1]`` comes from the ``y_coord`` and ``neighbor_coords[:, :, :, 0]`` comes from the ``x_coord``.

However, the bug is probably this: https://github.com/TuSimple/centerformer/blob/5a949b88ed7bb15aafb39bf78c95f1452063ebea/det3d/models/necks/rpn_transformer.py#L244-L246
The coordinates should be clamped separately using their own size rather than just assuming ``H = W``.

I don't have a machine at hand to test. Can you check if changing this can fix your error?

**Comment by Castiel-Lee** [comment]
Hello,

I notice that, the bug reporting can disappear through _x_coor_ and _y_coor_ being clamped within [0, H] and [0, W]. But it could probably leave the fact behind that ```x_coor = order // W and y_coor = order % W```. I did a small test:
![image](https://user-images.githubusercontent.com/91303830/220735616-81625342-d908-456d-bfe6-7979ad2a17a9.png)
According to the picture, the revised version seems correctly make x_coor = 7 and y_coor =4. 

Could you think this over and recheck the implementation when you are available? Of course, if "transposing" or some other purposes exist, I would apologize for my misunderstanding and bothering you.

**Comment by edwardzhou130** [comment]
Oh, I see. So the ``x_coor`` and ``y_coor`` here do not mean the row and col indexes of this value in the 2D matrix (like [7,4]). Conversely, it means the indexes on the height dimension (``y_coor``) and width dimension (``x_coor``), which are common to describe the coordinates of a pixel in an image.

I will recheck the code just in case there are some other bugs.

**Comment by Castiel-Lee** [comment]
> Oh, I see. So the `x_coor` and `y_coor` here do not mean the row and col indexes of this value in the 2D matrix (like [7,4]). Conversely, it means the indexes on the height dimension (`y_coor`) and width dimension (`x_coor`), which are common to describe the coordinates of a pixel in an image.
> 
> I will recheck the code just in case there are some other bugs.

Since height dimension (`y_coor`) and width dimension (`x_coor`), in Line 469, it should be `torch.stack([y_coor, x_coor], dim=2)`, right?

**Comment by edwardzhou130** [comment]
The order of the ``x_coor`` and ``y_coor`` variables does not matter in this case, as long as you remember the correct dimension that each variable represents. The deformable attention layer also requires storing the reference point position in the same way.

**Comment by Castiel-Lee** [comment]
Oh, I see. I will go through it again and recheck it. Thank you so much for the clarification.

**Comment by edwardzhou130** [comment]
Great! If you have any further questions or concerns, feel free to reopen the issue.
